### ISSUE TITLE
Bug: Migration is not working (MM 7.4.0)

### ISSUE THREAD

**Issue by haivala** [mattermost/focalboard#4018]
## Steps to reproduce the behavior

1. Upgrade mattermost to from 7.3.0 to 7.4.0
2. Boards are not working

## Expected behavior

Broards should be working

## Edition and Platform

 - Edition:  Mattermost Docker Team edition 
 - Version: 7.4.0

## Additional context 
Log says
```
mattermost-postgres-1    | 2022-10-14 07:24:49.204 UTC [22] LOG:  database system was shut down at 2022-10-14 07:24:44 UTC
mattermost-postgres-1    | 2022-10-14 07:24:49.212 UTC [1] LOG:  database system is ready to accept connections
mattermost-postgres-1    | 2022-10-14 07:24:55.131 UTC [40] ERROR:  column "dirty" does not exist at character 17
mattermost-postgres-1    | 2022-10-14 07:24:55.131 UTC [40] STATEMENT:  SELECT version, dirty FROM "focalboard_schema_migrations" LIMIT 1
mattermost-postgres-1    | 2022-10-14 07:25:00.711 UTC [41] ERROR:  cannot delete from scalar
mattermost-postgres-1    | 2022-10-14 07:25:00.711 UTC [41] STATEMENT:
mattermost-postgres-1    | 	        INSERT INTO Preferences (UserId, Category, Name, Value) SELECT Id, 'focalboard', 'welcomePageViewed', replace((Props->'focalboard_welcomePageViewed')::varchar, '"', '') FROM Users WHERE Props->'focalboard_welcomePageViewed' IS NOT NULL ON CONFLICT DO NOTHING;
mattermost-postgres-1    | 	        INSERT INTO Preferences (UserId, Category, Name, Value) SELECT Id, 'focalboard', 'hiddenBoardIDs', replace(replace(replace((Props->'hiddenBoardIDs')::varchar, '"[', '['), ']"', ']'), '\"', '"') FROM Users WHERE Props->'hiddenBoardIDs' IS NOT NULL ON CONFLICT DO NOTHING;
mattermost-postgres-1    | 	        INSERT INTO Preferences (UserId, Category, Name, Value) SELECT Id, 'focalboard', 'tourCategory', replace((Props->'focalboard_tourCategory')::varchar, '"', '') FROM Users WHERE Props->'focalboard_tourCategory' IS NOT NULL ON CONFLICT DO NOTHING;
mattermost-postgres-1    | 	        INSERT INTO Preferences (UserId, Category, Name, Value) SELECT Id, 'focalboard', 'onboardingTourStep', replace((Props->'focalboard_onboardingTourStep')::varchar, '"', '') FROM Users WHERE Props->'focalboard_onboardingTourStep' IS NOT NULL ON CONFLICT DO NOTHING;
mattermost-postgres-1    | 	        INSERT INTO Preferences (UserId, Category, Name, Value) SELECT Id, 'focalboard', 'onboardingTourStarted', replace((Props->'focalboard_onboardingTourStarted')::varchar, '"', '') FROM Users WHERE Props->'focalboard_onboardingTourStarted' IS NOT NULL ON CONFLICT DO NOTHING;
mattermost-postgres-1    | 	        INSERT INTO Preferences (UserId, Category, Name, Value) SELECT Id, 'focalboard', 'version72MessageCanceled', replace((Props->'focalboard_version72MessageCanceled')::varchar, '"', '') FROM Users WHERE Props->'focalboard_version72MessageCanceled' IS NOT NULL ON CONFLICT DO NOTHING;
mattermost-postgres-1    | 	        INSERT INTO Preferences (UserId, Category, Name, Value) SELECT Id, 'focalboard', 'lastWelcomeVersion', replace((Props->'focalboard_lastWelcomeVersion')::varchar, '"', '') FROM Users WHERE Props->'focalboard_lastWelcomeVersion' IS NOT NULL ON CONFLICT DO NOTHING;
mattermost-postgres-1    | 	
mattermost-postgres-1    | 	        UPDATE Users SET props = (props - 'focalboard_welcomePageViewed' - 'hiddenBoardIDs' - 'focalboard_tourCategory' - 'focalboard_onboardingTourStep' - 'focalboard_onboardingTourStarted' - 'focalboard_version72MessageCanceled' - 'focalboard_lastWelcomeVersion');
```

**Comment by haivala** [comment]
How to solve this?

**Comment by cobenash** [comment]
I faced the same situation. 

Try to reinstall but not working.

**Comment by haivala** [comment]
Are you telling me I need to wait a month to get the boards working again?

**Comment by cobenash** [comment]
> Are you telling me I need to wait a month to get the boards working again?

Is it possible to downgrade from 7.4 to 7.3 ?

I haven't try it yet. :)

**Comment by harshilsharma63** [comment]
@haivala @cobenash can you please share the list of tables you have in the DB and the schema of `focalboard_schema_migrations` table?

**Comment by moetiker** [comment]
I faced the same situation. Upgraded from 7.3.0

**Comment by oetiker** [comment]
it seems that the props column in the users table does not work quite as expected:

```sql
select props - 'hiddenBoardIDs' from users;
```
does not work. it returns
```
ERROR:  cannot delete from scalar
```

this breaks the 

```sql
UPDATE Users SET props = (props - 'focalboard_welcomePageViewed' - 'hiddenBoardIDs' - 'focalboard_tourCategory' - 'focalboard_onboardingTourStep' - 'focalboard_onboardingTourStarted' - 'focalboard_version72MessageCanceled' - 'focalboard_lastWelcomeVersion');
```
command. And thus the migrations fail.

**Comment by oetiker** [comment]
further analyzing this, it turns out that the props column can contain the VALUE 'null' :) I guess this could be an artifact from a mysql to postgresql migration ...

**Comment by oetiker** [comment]
Runninng

```sql
update users set props = null where props::text = 'null';
```
fixes the users table

**Comment by oetiker** [comment]
Maybe an even more radical fix would be appropriate.

```sql
update users set props = '{}':json where props::text not like '{%';
```
Since MM seems to assume that the props column does contain a hash ...

**Comment by haivala** [comment]
> ```sql
> update users set props = null where props::text = 'null';
> ```

i ran this but there is still one error:
```
mattermost-postgres-1    | 2022-10-18 07:26:30.653 UTC [22] LOG:  database system was shut down at 2022-10-18 07:26:28 UTC
mattermost-postgres-1    | 2022-10-18 07:26:30.658 UTC [1] LOG:  database system is ready to accept connections
mattermost-postgres-1    | 2022-10-18 07:26:38.701 UTC [40] ERROR:  column "dirty" does not exist at character 17
mattermost-postgres-1    | 2022-10-18 07:26:38.701 UTC [40] STATEMENT:  SELECT version, dirty FROM "focalboard_schema_migrations" LIMIT 1
```

**Comment by oetiker** [comment]
@haivala not sure why you are getting this ... it seems there was a dirty column in the schema_migrations table at some point, and there is special code in mm to detect if it is still there or not ... this code seems to fail in your case, thinking that the column is still there while it is not actually present ...

**Comment by oetiker** [comment]
https://github.com/mattermost/focalboard/blob/6f0c34ddd1ae275986ab491a05ee4ba9c30ad0b1/server/services/store/sqlstore/schema_table_migration.go#L107-L133

**Comment by haivala** [comment]
I have been upgrading this instance since MM version 2 or 3, can't remember even so I'm not surprised there something extra over the years. I'll try that!

**Comment by haivala** [comment]
That sql gave me this:
```
 count
-------
     0
(1 row)
```

**Comment by oetiker** [comment]
hmmm that looks good ... could it be, that you have some old plugin copy somewhere ? Because in the current focalboard code the word 'dirty' only occurs in that schema table migration routine ... Maybe you have a second (old) focalboard pluggin kicking around ?

**Comment by harshilsharma63** [comment]
> Runninng
> 
> ```sql
> update users set props = null where props::text = 'null';
> ```
> 
> fixes the users table.
> 
> Now restart mattermost and the boards are back!

@oetiker thanks a ton for the analysisfor and posting the fix! We'll definitally handle this and look into other invalid states the data in the DB might be in.

**Comment by harshilsharma63** [comment]
@haivala you's is a different issue. A while back we switched DB migration tools and the new tool stores data in a different format in `focalboard_schema_migrations` table. We wreote a custom code to transform this table's schema and the stored stored in it from the old format to the new format. It looks like that code is not doing its job perfectly, and this causing your issue.

Can you please share all tables in your DB and the schema of `focalboard_schema_migrations` table?

**Comment by haivala** [comment]
> hmmm that looks good ... could it be, that you have some old plugin copy somewhere ? Because in the current focalboard code the word 'dirty' only occurs in that schema table migration routine ... Maybe you have a second (old) focalboard pluggin kicking around ?

Any tips how to check that? I just recently moved from the old docker install to the new one. `volumes/app/mattermost/plugins` has `focalboard` folder.

**Comment by haivala** [comment]
> @haivala you's is a different issue. A while back we switched DB migration tools and the new tool stores data in a different format in `focalboard_schema_migrations` table. We wreote a custom code to transform this table's schema and the stored stored in it from the old format to the new format. It looks like that code is not doing its job perfectly, and this causing your issue.
> 
> Can you please share all tables in your DB and the schema of `focalboard_schema_migrations` table?

Sure. Can you give me the commands? I'm more familiar with Mysql.

**Comment by harshilsharma63** [comment]
@haivala to list all tables-

```sql
show tables;
```

to show schema of a the table-

```sql
desc focalboard_schema_migrations;
```

**Comment by haivala** [comment]
My point was that I have Postgres db. I would have know the mysql commands from the memory :)

**Comment by haivala** [comment]
```
\d focalboard_schema_migrations
           Table "public.focalboard_schema_migrations"
 Column  |         Type          | Collation | Nullable | Default
---------+-----------------------+-----------+----------+---------
 version | bigint                |           | not null |
 name    | character varying(64) |           | not null |
Indexes:
    "focalboard_temp_schema_migration_pkey" PRIMARY KEY, btree (version)
```

**Comment by harshilsharma63** [comment]
Can you also check the schema of `focalboard_temp_schema_migration` table?

**Comment by haivala** [comment]
```
                     List of relations
 Schema |               Name               | Type  | Owner
--------+----------------------------------+-------+--------
 public | audits                           | table | **
 public | bots                             | table | **
 public | channelmemberhistory             | table | **
 public | channelmembers                   | table | **
 public | channels                         | table | **
 public | clusterdiscovery                 | table | **
 public | commands                         | table | **
 public | commandwebhooks                  | table | **
 public | compliances                      | table | **
 public | db_lock                          | table | **
 public | db_migrations                    | table | **
 public | emoji                            | table | **
 public | fileinfo                         | table | **
 public | focalboard_blocks                | table | **
 public | focalboard_blocks_history        | table | **
 public | focalboard_board_members         | table | **
 public | focalboard_board_members_history | table | **
 public | focalboard_boards                | table | **
 public | focalboard_boards_history        | table | **
 public | focalboard_categories            | table | **
 public | focalboard_category_boards       | table | **
 public | focalboard_file_info             | table | **
 public | focalboard_notification_hints    | table | **
 public | focalboard_preferences           | table | **
 public | focalboard_schema_migrations     | table | **
 public | focalboard_sessions              | table | **
 public | focalboard_sharing               | table | **
 public | focalboard_subscriptions         | table | **
 public | focalboard_system_settings       | table | **
 public | focalboard_teams                 | table | **
 public | focalboard_users                 | table | **
 public | groupchannels                    | table | **
 public | groupmembers                     | table | **
 public | groupteams                       | table | **
 public | incomingwebhooks                 | table | **
 public | ir_category                      | table | **
 public | ir_category_item                 | table | **
 public | ir_channelaction                 | table | **
 public | ir_db_migrations                 | table | **
 public | ir_incident                      | table | **
 public | ir_metric                        | table | **
 public | ir_metricconfig                  | table | **
 public | ir_playbook                      | table | **
 public | ir_playbookautofollow            | table | **
 public | ir_playbookmember                | table | **
 public | ir_run_participants              | table | **
 public | ir_statusposts                   | table | **
 public | ir_system                        | table | **
 public | ir_timelineevent                 | table | **
 public | ir_userinfo                      | table | **
 public | ir_viewedchannel                 | table | **
 public | jobs                             | table | **
 public | licenses                         | table | **
 public | linkmetadata                     | table | **
 public | notifyadmin                      | table | **
 public | oauthaccessdata                  | table | **
 public | oauthapps                        | table | **
 public | oauthauthdata                    | table | **
 public | outgoingwebhooks                 | table | **
 public | pluginkeyvaluestore              | table | **
 public | postreminders                    | table | **
 public | posts                            | table | **
 public | preferences                      | table | **
 public | productnoticeviewstate           | table | **
 public | publicchannels                   | table | **
 public | reactions                        | table | **
 public | recentsearches                   | table | **
 public | remoteclusters                   | table | **
 public | retentionpolicies                | table | **
 public | retentionpolicieschannels        | table | **
 public | retentionpoliciesteams           | table | **
 public | roles                            | table | **
 public | schema_migrations                | table | **
 public | schemes                          | table | **
 public | sessions                         | table | **
 public | sharedchannelattachments         | table | **
 public | sharedchannelremotes             | table | **
 public | sharedchannels                   | table | **
 public | sharedchannelusers               | table | **
 public | sidebarcategories                | table | **
 public | sidebarchannels                  | table | **
 public | status                           | table | **
 public | systems                          | table | **
 public | teammembers                      | table | **
 public | teams                            | table | **
 public | termsofservice                   | table | **
 public | threadmemberships                | table | **
 public | threads                          | table | **
 public | tokens                           | table | **
 public | uploadsessions                   | table | **
 public | useraccesstokens                 | table | **
 public | usergroups                       | table | **
 public | users                            | table | **
 public | usertermsofservice               | table | **
 ```

**Comment by haivala** [comment]
> focalboard_temp_schema_migration

\d focalboard_temp_schema_migration
Did not find any relation named "focalboard_temp_schema_migration".

**Comment by harshilsharma63** [comment]
@haivala okay, I feel I know what's hapenning. Can you show the contents of `focalboard_system_settings` table? Also, to summarise, whats the latest error you get on starting Focalboard?

**Comment by cobenash** [comment]
> focalboard_system_settings

Here is mine. @harshilsharma63 

```

mattermost=# select * from focalboard_system_settings;
                    id                    |                value
------------------------------------------+--------------------------------------
 TelemetryID                              | f6357809-a052-4f0c-85e4-5d6c92a287e0
 UniqueIDsMigrationComplete               | true
 TeamLessBoardsMigrationComplete          | true
 DeletedMembershipBoardsMigrationComplete | true
 CategoryUuidIdMigrationComplete          | true
(5 rows)
```

I have the same error message as @haivala

**Comment by haivala** [comment]
> select * from focalboard_system_settings;

I have the same

**Comment by harshilsharma63** [comment]
@haivala can you please double check that you are indeed runnning focalboard version 7.4? Try removing and re-installing the plugin. The DB schema checks out - no `dirty` column in `focalboard_schema_migrations` is correct, but if its still trying to do `SELECT version, dirty FROM "focalboard_schema_migrations" LIMIT 1`, that occurred in an older version of Focalboard.

Feel free to DM me on community.mattermost.com/ (username: `harshil.sharma`) and I'll look into your issue on your system if you;'re fine with it

**Comment by harshilsharma63** [comment]
@haivala no, it doesn't delete the data.

**Comment by haivala** [comment]
So in system console: disable, remove and then in marketplace install?

**Comment by harshilsharma63** [comment]
Disable from system console, remove, and install [from here](https://github.com/mattermost/focalboard/releases/tag/v7.4.2) just to be extra sure of the version we're instaling.

**Comment by cobenash** [comment]
> Disable from system console, remove, and install [from here](https://github.com/mattermost/focalboard/releases/tag/v7.4.2) just to be extra sure of the version we're instaling.

@harshilsharma63 How could we install [from here](https://github.com/mattermost/focalboard/releases/tag/v7.4.2) when we use docker image as the main installation method?

**Comment by cobenash** [comment]
> > Disable from system console, remove, and install [from here](https://github.com/mattermost/focalboard/releases/tag/v7.4.2) just to be extra sure of the version we're instaling.
> 
> @harshilsharma63 How could we install [from here](https://github.com/mattermost/focalboard/releases/tag/v7.4.2) when we use docker image as the main installation method?

I got it.

I download it  via following command 
```
wget https://github.com/mattermost/focalboard/releases/download/v7.4.2/mattermost-plugin-focalboard.tar.gz
```
And I uncompress the file and then put it to the plugin folder. The focalboard function is back.

**Comment by haivala** [comment]
did you extract it in `volumes/app/mattermost/plugins/focalboard` or in `volumes/app/mattermost/client/plugins/focalboard`?
Which one is the right one? Do I need them both?

**Comment by cobenash** [comment]
> did you extract it in `volumes/app/mattermost/plugins/focalboard` or in `volumes/app/mattermost/client/plugins/focalboard`? Which one is the right one? Do I need them both?

`volumes/app/mattermost/plugins/focalboard`

All the board including data is back too. :)

Thanks for everyone's help.

Appreciate it.

**Comment by haivala** [comment]
@cobenash Do you have them both?

**Comment by harshilsharma63** [comment]
@cobenash I'm super glad it worked for you!

@haivala were you able to get it to work?

**Comment by haivala** [comment]
> @cobenash I'm super glad it worked for you!
> 
> @haivala were you able to get it to work?

I'm waiting answer to: Do I need them both?

**Comment by harshilsharma63** [comment]
@haivala its perfectly safe to do both. Your data is never deleted.

**Comment by haivala** [comment]
> @haivala its perfectly safe to do both. Your data is never deleted.

I mean is it right to have the plugin in both folders? or should the client side one be deleted?

**Comment by harshilsharma63** [comment]
@haivala I'm sorry but I have never used Mattermost with Docker so I'm not sure on this. If you just go into system console > plugins, do you get an option to remove Focalboard and an upload plugin button?

**Comment by haivala** [comment]
I can now confirm that the error has vanished after plugin replace.

Thanks for everyone's help!

**Comment by cobenash** [comment]
@harshilsharma63  Thanks 💯 

@haivala I am glad that you worked too. I replied via modifying the same comment. lol

`/volumes/app/mattermost/client/plugins/*` looks like the location to store user sensitive information compiled by the plugin in `/volumes/app/mattermost/plugins/*`

**Comment by harshilsharma63** [comment]
Closing as fixed by #4029
